src/compute-client/src/plan/transform/relax_must_consolidate.rs

```diff
@@ -9,3 +9,74 @@
 
 //! Placeholder module for an [crate::plan::transform::Transform] that infers
 //! physical monotonicity.
+
+use std::marker::PhantomData;
+
+use crate::plan::interpret::{PhysicallyMonotonic, SingleTimeMonotonic};
+use crate::plan::reduce::{HierarchicalPlan, MonotonicPlan};
+use crate::plan::top_k::{MonotonicTop1Plan, MonotonicTopKPlan, TopKPlan};
+use crate::plan::transform::{BottomUpTransform, TransformConfig};
+use crate::plan::{Plan, ReducePlan};
+
+/// A transformation that takes the result of single-time physical monotonicity
+/// analysis and refines, as appropriate, the setting of the `must_consolidate`
+/// flag in monotonic `Plan` nodes with forced consolidation.
+#[derive(Debug)]
+pub struct RelaxMustConsolidate<T = mz_repr::Timestamp> {
+    _phantom: PhantomData<T>,
+}
+
+impl<T> RelaxMustConsolidate<T> {
+    pub fn new() -> Self {
+        RelaxMustConsolidate {
+            _phantom: Default::default(),
+        }
+    }
+}
+
+impl<'a, T> BottomUpTransform<'a, T> for RelaxMustConsolidate<T> {
+    type Info = PhysicallyMonotonic;
+
+    type Interpreter = SingleTimeMonotonic<'a, T>;
+
+    fn name(&self) -> &'static str {
+        "must_consolidate relaxation"
+    }
+
+    fn interpreter(config: &'a TransformConfig) -> Self::Interpreter {
+        SingleTimeMonotonic::new(&config.monotonic_ids)
+    }
+
+    fn action(plan: &mut Plan<T>, _plan_info: &Self::Info, input_infos: &[Self::Info]) {
+        // Look at `input_infos` and type of `Plan` node and refine the `must_consolidate` flag.
+        // Here, we only care about LIR nodes that have a single input.
+        match (plan, input_infos.first()) {
```

Comment:
nit: I think this should work with just matching on `input_infos` here and using `[ PhysicallyMonotonic(true) ]` in the pattern below because all plans matched inside have a single input, but I'll leave the final decision to you.

Reply:
It does! Happy to go with matching the slice.